App_Use_Cases.tex

```diff
@@ -0,0 +1,447 @@
+
+When initializing PMIx, programming models can register themselves, including their name, version, and threading model.  This information is then cached locally and can then be read asynchronously by other programming models using PMIx's Event Notification system (see next section for more details).
+
+This initialization mechanism also allows libraries to share knowledge of each other's resources and intended resource utilization. For example, if OpenMP knows which hardware threads that MPI is using it could potentially avoid processor and cache contention.
+
+\littleheader{Code Example}
+
+\pmixCodeImportC[]{sources/_autogen_/hybrid-prog-model.c_declare_model}
+
+\littleheader{Related Interfaces}
+
+{\large \refapi{PMIx_Init}}
+\pasteSignature{PMIx_Init}
+
+\littleheader{Related Attributes}
+
+\pasteAttributeItem{PMIX_PROGRAMMING_MODEL}
+\pasteAttributeItem{PMIX_MODEL_LIBRARY_NAME}
+\pasteAttributeItem{PMIX_MODEL_LIBRARY_VERSION}
+\pasteAttributeItem{PMIX_THREADING_MODEL}
+\pasteAttributeItem{PMIX_MODEL_NUM_THREADS}
+\pasteAttributeItem{PMIX_MODEL_NUM_CPUS}
+\pasteAttributeItem{PMIX_MODEL_CPU_TYPE}
+\pasteAttributeItem{PMIX_MODEL_PHASE_NAME}
+\pasteAttributeItem{PMIX_MODEL_PHASE_TYPE}
+\pasteAttributeItem{PMIX_MODEL_AFFINITY_POLICY}
+
+\subsubsection{Coordinating at Runtime}
+
+The PMIx Event Notification system provides a mechanism by which the resource manager can communicate system events to applications, thus providing applications with an opportunity to generate an appropriate response. Hybrid applications can leverage these events for cross-library coordination.
+
+Programming models can access the information provided by other programming models during their initialization using the event notification system.  In this case, programming models should register a callback for the \refconst{PMIX_MODEL_DECLARED} event.
+
+Programming models can also use the PMIx event notification system to communicate dynamic information, such as entering a new application phase (\pasteAttributeItem{PMIX_MODEL_PHASE_NAME}) or a change in resources used (\refconst{PMIX_MODEL_RESOURCES}).  This dynamic information can be broadcast to other programming models using the \refapi{PMIx_Notify_event} function.  Other programming models can register callback functions to run when these events occur (i.e., callback functions) using \refapi{PMIx_Register_event_handler}.
```

Comment:
Runtime libraries typically cannot know about application phases

Reply:
Good point.  I also don't think it will necessarily be runtime libraries that are doing the notification of resource changes (that's probably more the domain of resource managers).  What are your thoughts on:

```suggestion
Applications, runtime libraries, and resource managers can also use the PMIx event notification system to communicate dynamic information, such as entering a new application phase (\pasteAttributeItem{PMIX_MODEL_PHASE_NAME}) or a change in resources used (\refconst{PMIX_MODEL_RESOURCES}). This dynamic information can be broadcast using the \refapi{PMIx_Notify_event} function.  Runtime libraries can register callback functions to run when these events occur using \refapi{PMIx_Register_event_handler}.
```